frame/dvm/src/lib.rs

```diff
@@ -501,15 +413,36 @@ impl<T: Config> Pallet<T> {
 			transaction.source,
 			contract_address.unwrap_or_default(),
 			transaction_hash,
-			reason,
+			reason.clone(),
 		));
-		Ok(PostDispatchInfo {
-			actual_weight: Some(T::GasWeightMapping::gas_to_weight(
-				used_gas.unique_saturated_into(),
-			)),
-			pays_fee: Pays::No,
-		})
-		.into()
+		Ok((reason, used_gas))
+	}
+
+	/// Pure read-only call to contract, the sender is INTERNAL_CALLER
+	/// NOTE: You should never use raw call for any non-read-only operation. Because raw_call
+	/// has ability to chang storage. Be carefully.
+	pub fn raw_call(contract: H160, input: Vec<u8>) -> Result<Vec<u8>, DispatchError> {
+		let (_, _, info) = Self::execute(
+			INTERNAL_CALLER,
+			input,
+			U256::zero(),
+			U256::from(INTERNAL_TX_GAS_LIMIT),
+			None,
+			None,
+			TransactionAction::Call(contract),
+			None,
+		)?;
```

Comment:
Could we make it readonly?

Reply:
This is not possible at the moment, the call here will directly call the executor in evm. Now, I added some docs here to avoid `raw_call` abuse.

Reply:
Try this way, and add a unit test.
```
		let info = with_transaction(|| {
			let (_, _, info) = Self::execute(
				INTERNAL_CALLER,
				input.clone(),
				U256::zero(),
				U256::from(0x300000),
				None,
				None,
				TransactionAction::Call(contract),
				None,
			)?;

			Rollback(info)
		});
```
cc @AurevoirXavier @xiaoch05

Reply:
Fixed